cEP-0018.md

```diff
@@ -272,3 +272,157 @@ class TestBear(ASTBear):
+result are not returned when simply fetching text from a node in parse tree.
+
+Consider the code:
+
+```python
+def do_nothing():
+    pass
+
+def run_coala(console_printer=None,
+              log_printer=None,
+              print_results=do_nothing,
+              print_section_beginning=do_nothing,
+              nothing_done=do_nothing,
+              autoapply=True,
+              force_show_patch=False,
+              arg_parser=None,
+              arg_list=None,
+              args=None,
+              debug=False):
+    """
+    This function signature has been taken from coala-core for fairly complex
+    example.
+    """
+    print('This function will do something')
+    if (console_printer and
+        log_printer and
+        not (args == None and debug == False)
+        ):
+       print('These conditions were made to trick !')
+       print(' !(a & b) == !a or !b')
+```
+
+The corresponding parse-tree:
+![example_tree_complex](https://user-images.githubusercontent.com/10978108/43359512-e38dbfa0-92c0-11e8-8f5b-fab244d27ec0.png)
+
+After converting this source back using:
+
+```python
+#!/usr/bin/env python
+from antlr4 import *
+from Python3Lexer import Python3Lexer
+from Python3Parser import Python3Parser
+from Python3Visitor import Python3Visitor
+
+import sys
+
+class Visitor(Python3Visitor):
+    def visitFile_input(self, ctx):
+        print(ctx.getText())
+
+def main(args):
+    file = FileStream(args[1])
+    lexer = Python3Lexer(file)
+    stream = CommonTokenStream(lexer)
+    parser = Python3Parser(stream)
+
+    tre = parser.file_input()
+    if parser.getNumberOfSyntaxErrors()!=0:
+        print("File contains {} "
+              "syntax errors".format(parser.getNumberOfSyntaxErrors()))
+        return
+
+    visitor = Visitor()
+    visitor.stream = stream
+    visitor.visit(tre)
+
+if __name__ == '__main__':
+    main(sys.argv)
+```
+
+We get the following output:
+
+```python
+defdo_nothing():
+    pass
+
+defrun_coala(console_printer=None,log_printer=None,print_results=do_nothing,print_section_beginning=do_nothing,nothing_done=do_nothing,autoapply=True,force_show_patch=False,arg_parser=None,arg_list=None,args=None,debug=False):
+    """
+    This function signature has been taken from coala-core for fairly complex
+    example.
+    """
+print('This function will do something')
+if(console_printerandlog_printerandnot(args==Noneanddebug==False)):
+       print('These conditions were made to trick !')
+print(' !(a & b) == !a or !b')
+
+
+<EOF>
+```
+
+The most important points to note in the above are (Especially with respect
+to a `Python` grammar scenario):
+
+- All spaces within two tokens has been removed unless present inside a string
+- All newlines have been stripped unless the newline is specified by the
+  grammar to be present in the parse tree. With reference to the Python grammar
+  the newlines present in irregularly because they are preserved only in case
+  of the `compund_stmt` rule.
```

Comment:
-> compound ?

Reply:
I was trying to refer to [this](https://github.com/antlr/grammars-v4/blob/master/python3-py/Python3.g4#L154) rule :sweat_smile: